sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,843 @@
+        A dictionary of information created by the L-BFGS optimizer during
+        fitting.
+
+
+    Examples
+    --------
+    >>> X = [[0], [1], [2], [3]]
+    >>> y = [0, 0, 1, 1]
+    >>> from sklearn.neighbors import LargeMarginNearestNeighbor
+    >>> lmnn = LargeMarginNearestNeighbor(n_neighbors=1)
+    >>> lmnn.fit(X, y) # doctest: +ELLIPSIS
+    LargeMarginNearestNeighbor(...)
+    >>> print(lmnn.predict([[1.1]]))
+    [0]
+    >>> print(lmnn.predict_proba([[0.9]]))
+    [[ 1.  0.]]
+
+
+    References
+    ----------
+    .. [1] `Weinberger, K. Q.; Saul L. K. (2009). "Distance Metric Learning
+    for Large Margin Classification". Journal of Machine Learning Research.
+    10: 207–244.
+    <http://jmlr.csail.mit.edu/papers/volume10/weinberger09a/weinberger09a.pdf>`_
+
+    .. [2] `Wikipedia entry on Large Margin Nearest Neighbor
+           <https://en.wikipedia.org/wiki/Large_margin_nearest_neighbor>`_
+
+    """
+
+    def __init__(self, L=None, n_neighbors=3, n_features_out=None,
+                 max_iter=200, tol=1e-5, use_pca=True, max_constr=int(1e7),
+                 use_sparse=True, load=None, save=None, disp=0,
+                 random_state=None):
+
+        super(LargeMarginNearestNeighbor, self).__init__(
+            n_neighbors=n_neighbors)
+
+        # Parameters
+        self.L = L
+        self.n_features_out = n_features_out
+        self.max_iter = max_iter
+        self.tol = tol
+        self.use_pca = use_pca
+        self.max_constr = max_constr
+        self.use_sparse = use_sparse
+        self.load = load
+        self.save = save
+        self.disp = disp
+        self.random_state = random_state
+
+    def fit(self, X, y):
+        """Find a linear transformation by optimization of the unconstrained
+        problem, such that the k-nearest neighbor classification accuracy
+        improves.
+
+        Parameters
+        ----------
+        X : array-like
+            An array of training samples with shape (n_samples, n_features_in).
+        y : array-like
+            An array of data labels with shape (n_samples,).
+
+        Returns
+        -------
+        LargeMarginNearestNeighbor
+            self
+
+        """
+
+        # Check inputs consistency
+        self.X_, y = check_X_y(X, y, order='F')
+        check_classification_targets(y)
+
+        # Store the appearing classes and the class index for each sample
+        self.labels_, self.y_ = np.unique(y, return_inverse=True)
+        self.classes_ = np.arange(len(self.labels_))
+
+        # Check that the number of neighbors is achievable for all classes
+        self.n_neighbors_ = self.check_n_neighbors(self.y_)
+        # TODO: Notify superclass KNeighborsClassifier that n_neighbors
+        # might have changed to n_neighbors_
+        # super().set_params(n_neighbors=self.n_neighbors_)
+
+        # Initialize transformer
+        self.L_, self.n_features_out_ = self._init_transformer()
+
+        # Prepare for saving if needed
```

Comment:
Hi, thanks for the comments. I put this in case the training takes a long time and one wants to abruptly stop it without losing the intermediate solution. Yes, I think a `warm_start` would suffice to replace `load`. I just saw how `MLPClassifier` does it. I can remove `save` completely.